src/ddscxx/include/org/eclipse/cyclonedds/topic/byteswapping.hpp

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright(c) 2006 to 2020 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+
+#ifndef IDL_BYTESWAPPING_HPP
+#define IDL_BYTESWAPPING_HPP
+
+namespace org
+{
+  namespace eclipse
+  {
+    namespace cyclonedds
+    {
+      namespace topic
+      {
+
+        inline void bswap_16(void* in)
```

Comment:
You should have checked the godbolt results for these kinds of operations generated by MSVC, not impressive at all.

I like the approach of using templates and selective generation here, very nice! Though the type punning is still happening "under the hood" inside the union.

I don't understand the risk from arithmetic types, byteswapping something which has a sizeof which is not a power of 2 is meaningless, right?

Using references in stead of return types should not be an issue, since the compiler should be able to determine that taking something as a parameter and then writing the result back to the same variable would be compiled as using a reference.

Reply:
> Though the type punning is still happening "under the hood" inside the union.

You're correct, of course: this, too, is type punning. The difference is that (at least in my understanding of the intricacies), the version you proposed is undefined behaviour, whereas the variant I suggested is defined *as an implementation-specific extension of gcc* (when compiling C, presumably also for C++; and presumably also present in clang). That italicised bit I forgot about ... Moreover, there is a nagging worry in the back of my mind that unions in C++ are more restrictive than in C. So what the C++ rules are exactly, I don't really know.

I think that means the one standard-sanctioned exception then is accessing the bytes only via a `char *` or `memcpy`/`memmove`. (I am positive those are allowed, the "I think" concerns the absence of other means.)

So memcpy'ing the data into a local variable, then memcpy it back is an option, casting the pointer to a `char *` and doing a byte-swapping loop is one, doing the byteswapping in-situ in the CDR before extracting the data (it is an array of bytes anyway! and that way you'll never need to consider endianness of it afterward, which is what the C type support in Cyclone currently does).

> You should have checked the godbolt results for these kinds of operations generated by MSVC, not impressive at all.

Fortunately gcc and clang are onto all of these tricks and turn it into efficient machine code (that's *including* 
```C++
    unsigned char *z = static_cast<unsigned char *>(in);
    int i = 0, j = 7;
    while(i<j) {
        unsigned char t = z[i];
        z[i] = z[j];
        z[j] = t;
    }
```
😄

As to MS's compiler ... if a platform's native compiler suite is a second-rate compiler (and it arguably is, considering its limitations), surely that's a second-rate platform that deserves second-rate performance? If there is a way to do it that is handled properly by that compiler (https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/byteswap-uint64-byteswap-ulong-byteswap-ushort?view=msvc-160 perhaps) then that it is of course best.

> I don't understand the risk from arithmetic types, byteswapping something which has a sizeof which is not a power of 2 is meaningless, right?

Well ... if I were to build a machine with where the size of a word equalled 3 addressing units ... I would still have endianness to worry about! And those did exist 😄 https://en.wikipedia.org/wiki/Gemini_Guidance_Computer  (https://en.wikipedia.org/wiki/Electrologica_X8 was word-addressable or would've used that; see https://en.wikipedia.org/wiki/Word_(computer_architecture) for what might have been).

Evidently, that machine is not really something we need to be concerned about. What prompted my remark is that 128-bit integers and floating-point numbers do exist out there, and that consequently it is not very wise to assume that all integral and floating-point types have sizes ≤ 8 units.